File: app/javascript/channels/tweet_channel.js
import consumer from "./consumer"


$(function () {

  const commentArea = document.getElementById("comment-area")

  if (commentArea) {
    const tweetChannel = consumer.subscriptions.create({
      channel: 'TweetChannel',
      tweet: $('#comments').data('tweet_id')
    }, {
      connected() {
        // Called when the subscription is ready for use on the server
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      received: function (data) {
        return $('#comments').append(data['comment']);
      },

      speak: function (comment, mensionUserId) {
        return this.perform('speak', {
          comment: comment,
          mension_user_id: mensionUserId
        });
      }
    });

    const commentButton = document.getElementById("comment-button")
    commentButton.addEventListener("click", function () {
      let content = commentArea.value
      let mensionUserId = commentArea.dataset.mension_user_id
      tweetChannel.speak(content, mensionUserId)
      commentArea.value = ''
      mention.value = ''
      commentArea.dataset.mension_user_id = ''
    })
  }
});

File: app/channels/tweet_channel.rb
class TweetChannel < ApplicationCable::Channel
  def subscribed
    stream_from "tweet_channel_#{params["tweet"]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    current_user.comments.create!(
      content: data["comment"],
      tweet_id: params["tweet"],
      mention_user_id: data["mention_user_id"],
    )
  end
end
